### ISSUE TITLE
New nuclear feature: podcast feature

### ISSUE THREAD

**Issue by Architsakri** [nukeop/nuclear#966]
using the apple podcast api, we can import podcasts for free an play them on nuclear

**Comment by Ivan12273** [comment]
Hi, I'm new collaborating on OP projects, is anyone already working on this? I would like to help with this new feature :)

**Comment by nukeop** [comment]
As far as I know, nobody picked it up yet. Would you like to be assigned? Not sure if we need any new components designed?

**Comment by Ivan12273** [comment]
Yes, I would like it. I saw that the API requests are in core > rest. 

![image](https://user-images.githubusercontent.com/42753018/121740120-3b29e680-cac2-11eb-922c-8460f43a27f7.png)

A new item called "podcast" next to the live broadcast could be?

**Comment by nukeop** [comment]
Sure, I think that's a good way to handle it. You are right that it should be placed in core.
Here's some more info on what should go where: https://app.gitbook.com/@nukeop/s/nuclear/developer-resources/architecture

Let me know if you have any questions.

**Comment by Ivan12273** [comment]
Alright, thanks :)

**Comment by Architsakri** [comment]
wait hey, since last.fm's api can pick up most played songs, artists, and all that. We should change the dashboard and add a recommendation engine so we can have recommended artists, most played songs and artists and all that.

**Comment by nukeop** [comment]
There's autoradio already, but we don't keep a history of played tracks.

**Comment by Architsakri** [comment]
I was thinking that if this had a mobile version it would really blow up. Is there any chance of a mobile version?

**Comment by nukeop** [comment]
I tried developing on an Android VM but it's too slow and unreliable. Iphone version is not feasible of course.

**Comment by Architsakri** [comment]
yeah I am in the process of developing with react native. So far I have made a login system, and a dashboard synced with last.fm just to get information about the current trend. Would you want to test the android apk?

**Comment by Architsakri** [comment]
but leaving that aside, can we be expecting a feature where you can cast your audio to speakers right from the app?

**Comment by nukeop** [comment]
Using chromecast?

**Comment by Architsakri** [comment]
yeah

**Comment by Ivan12273** [comment]
Hello, I did not have problems creating a new plugin for search podcast with the iTunes API.
But I had troubles doing the meta provider. I can receive all the podcast: 

![image](https://user-images.githubusercontent.com/42753018/122283257-fa5f1280-ceb1-11eb-8f17-fed0807e70d6.png)

But when I click on one of the podcast I am not receiving the artistId in the function fetchArtistDetails. I 
I am planning doing it in a new tab called "podcast" but first I am trying to doing it on the "ARTISTS" tab, but I am having this problem, I am missing something?

This is my code:

`import MetaProvider from '../metaProvider';
import * as iTunes from '../../rest/iTunesPodcast';
import { 
  SearchResultsArtist, 
  SearchResultsAlbum, 
  SearchResultsTrack, 
  ArtistDetails, 
  AlbumDetails,
  SearchResultsSource 
} from '../plugins.types';
import LastFmApi from '../../rest/Lastfm';

class iTunespodcastMetaProvider extends MetaProvider {
  lastfm: LastFmApi;
  constructor() {
    super();
    this.name = 'iTunespodcast Meta Provider';
    this.sourceName = 'iTunespodcast Meta Provider';
    this.description = 'Metadata provider that uses iTunes as a source.';
    this.searchName = 'iTunespodcast';
    this.image = null;
    this.lastfm = new LastFmApi(process.env.LAST_FM_API_KEY, process.env.LASTFM_API_SECRET);
  }

  searchForArtists(query: string): Promise<Array<SearchResultsArtist>> {
    return iTunes.podcastSearch(query, '50')
      .then(response => response.json())
      .then(response => response.results.map(podcast => ({
        id: `${podcast.collectionId}`,
        coverImage: '',
        thumb: '',
        name: podcast.trackName,
        source: SearchResultsSource.iTunesPodcast
      })));
  }

  async searchForReleases(query: string): Promise<SearchResultsAlbum[]> {
    await this.searchForTracks(query);
    return Promise.resolve([]);
  }

  searchForTracks(query: string): Promise<SearchResultsTrack[]> {
    return iTunes.podcastSearch(query, '50')
      .then(response => response.json())
      .then((json) => json.results.map(podcast => ({
        id: podcast.collectionId,
        title: podcast.collectionName,
        artist: podcast.collectionName,
        source: SearchResultsSource.iTunesPodcast
      })));
  }

  async searchAll(query: string): Promise<{ 
    artists: SearchResultsArtist[]; 
    releases: SearchResultsAlbum[]; 
    tracks: SearchResultsTrack[]; 
  }> {
    const artists = await this.searchForArtists(query);
    const releases = await this.searchForReleases(query);
    const tracks = await this.searchForTracks(query);
    return Promise.resolve({ artists, releases, tracks });
  }

  async fetchArtistDetails(artistId: string): Promise<ArtistDetails> {
    console.log(artistId);
  }

 /* other functions */
 
}

export default iTunespodcastMetaProvider;
`

The name of the file is `iTunespodcast.ts` I export it on the index file as:
`export { default as iTunespodcastMetaProvider } from './iTunespodcast';`

**Comment by Ivan12273** [comment]
Hello, I did not have problems creating a new plugin for search podcast with the iTunes API.
But I had troubles doing the meta provider. I can receive all the podcast: 

![image](https://user-images.githubusercontent.com/42753018/122283257-fa5f1280-ceb1-11eb-8f17-fed0807e70d6.png)

But when I click on one of the podcast I am not receiving the artistId in the function fetchArtistDetails. I 
I am planning doing it in a new tab called "podcast" but first I am trying to doing it on the "ARTISTS" tab, but I am having this problem, I am missing something?

This is my code:

```
import MetaProvider from '../metaProvider';
import * as iTunes from '../../rest/iTunesPodcast';
import { 
  SearchResultsArtist, 
  SearchResultsAlbum, 
  SearchResultsTrack, 
  ArtistDetails, 
  AlbumDetails,
  SearchResultsSource 
} from '../plugins.types';
import LastFmApi from '../../rest/Lastfm';

class iTunespodcastMetaProvider extends MetaProvider {
  lastfm: LastFmApi;
  constructor() {
    super();
    this.name = 'iTunespodcast Meta Provider';
    this.sourceName = 'iTunespodcast Meta Provider';
    this.description = 'Metadata provider that uses iTunes as a source.';
    this.searchName = 'iTunespodcast';
    this.image = null;
    this.lastfm = new LastFmApi(process.env.LAST_FM_API_KEY, process.env.LASTFM_API_SECRET);
  }

  searchForArtists(query: string): Promise<Array<SearchResultsArtist>> {
    return iTunes.podcastSearch(query, '50')
      .then(response => response.json())
      .then(response => response.results.map(podcast => ({
        id: `${podcast.collectionId}`,
        coverImage: '',
        thumb: '',
        name: podcast.trackName,
        source: SearchResultsSource.iTunesPodcast
      })));
  }

  async searchForReleases(query: string): Promise<SearchResultsAlbum[]> {
    await this.searchForTracks(query);
    return Promise.resolve([]);
  }

  searchForTracks(query: string): Promise<SearchResultsTrack[]> {
    return iTunes.podcastSearch(query, '50')
      .then(response => response.json())
      .then((json) => json.results.map(podcast => ({
        id: podcast.collectionId,
        title: podcast.collectionName,
        artist: podcast.collectionName,
        source: SearchResultsSource.iTunesPodcast
      })));
  }

  async searchAll(query: string): Promise<{ 
    artists: SearchResultsArtist[]; 
    releases: SearchResultsAlbum[]; 
    tracks: SearchResultsTrack[]; 
  }> {
    const artists = await this.searchForArtists(query);
    const releases = await this.searchForReleases(query);
    const tracks = await this.searchForTracks(query);
    return Promise.resolve({ artists, releases, tracks });
  }

  async fetchArtistDetails(artistId: string): Promise<ArtistDetails> {
    console.log(artistId);
  }

 /* other functions */
 
}

export default iTunespodcastMetaProvider;
```

The name of the file is `iTunespodcast.ts` I export it on the index file as:
`export { default as iTunespodcastMetaProvider } from './iTunespodcast';`

**Comment by Architsakri** [comment]
I remember there was a project called podrr and it used the iTunes api to get the top podcasts by country, get the album covers, etc. Refer to it for a better understanding if it helps: https://github.com/Sn8z/Poddr  :)

**Comment by Architsakri** [comment]
but as a suggestion cant we use Ethereum and the IPFS protocol to make the platform truly decentralized. With the larger streaming library and the use of Ethereum and IPFS it will be more attention grabbing.

**Comment by nukeop** [comment]
We will definitely not be attaching any crypto scams to this program.

**Comment by HaiDang666** [comment]
No idea why this app need decentralized

**Comment by nukeop** [comment]
The funny thing is, it's a desktop program so this concept doesn't even apply to it.

**Comment by nukeop** [comment]
@Ivan12273 Could you post this code in a draft pull request? It will be easier to read and analyze and I will be able to test it.

**Comment by Ivan12273** [comment]
@nukeop Of course, thank you :) 

https://github.com/nukeop/nuclear/pull/974

**Comment by HaiDang666** [comment]
resolves #974
